### PR TITLE
Cards in dashboards: milestone 1 (backend)

### PR DESCRIPTION
**Pull request by johnswanson**
- migrations to add `report_card.dashboard_id`, and an index/foreign key for it
- accept a `dashboard_id` parameter for `POST /api/card/`
- accept a `dashboard_id` parameter for `PUT /api/card/`
- return `dashboard_id` from the search API
- do not show dashboard questions without a DashboardCard in the search API
- do not show dashboard questions in the autocomplete API
- modifies `POST /api/dashboard/:id/copy` to always copy dashboard questions
- modifies `PUT /api/dashboard/:id` to automatically move dashboard questions along with the dashboard
- modifies the Search API to allow including (or not) dashboard questions

Product doc: https://www.notion.so/metabase/Enable-saving-questions-directly-to-dashboards-and-make-this-the-default-path-a84be6e0540241189b251401fb35ca91#f9bbd61d8bc442ab8eaf908c680394b4

Epic: https://github.com/metabase/metabase/issues/47326

### CODE REVIEW

File: resources/migrations/001_update_migrations.yaml [hunk 1]
```diff
@@ -9270,6 +9270,90 @@ databaseChangeLog:
                   type: ${text.type}
                   remarks: running, failed, or completed
 
+
+  - changeSet:
+      id: v51.2024-08-28T07:03:41
```
Comment (crisptrutski) [inline resources/migrations/001_update_migrations.yaml:9275]: Following the discussion in our last guild, should we not add a descriptive suffix to these ids?

File: resources/migrations/001_update_migrations.yaml [hunk 2]
```diff
@@ -9270,6 +9270,90 @@ databaseChangeLog:
...
+      changes:
+        - addColumn:
+            tableName: report_dashboardcard
+            columns:
+              - column:
+                  name: is_dashboard_internal_card
+                  type: ${boolean.type}
+                  remarks: |
+                    A denormalized representation that this edge from dashboard<->card should be the *only* edge from its card.
+                    Provides an efficient way to enforce the fact that
```
Comment (crisptrutski) [inline resources/migrations/001_update_migrations.yaml:9302]: Watch out for this trailing sentence. How about the following?

```suggestion
                    A denormalized edge from dashboard<->card which uses a unique index to ensure that
                    there is at most one `is_dashboard_internal_card` dashcard per card.
```

This also seems to be attached to the wrong changeset, and belongs on the next one.

File: resources/migrations/001_update_migrations.yaml [hunk 3]
```diff
@@ -9270,6 +9270,90 @@ databaseChangeLog:
...
+                  defaultValueBoolean: false
+                  constraints:
+                    nullable: false
+
+  - changeSet:
+      id: v51.2024-08-28T07:03:43
+      author: johnswanson
+      comment: Add `report_dashboardcard.dashboard_internal_card_id`
```
Comment (crisptrutski) [inline resources/migrations/001_update_migrations.yaml:9310]: This unique field won't guarantee that there aren't some non-`is_dashboard_internal_card` dashcards for the given card 🤔

**Review by crisptrutski — COMMENTED** [review]
~~Can you drop a link to the epic or project doc in the description? I'm curious to understand more about the internal card feature.~~ https://www.notion.so/metabase/Cards-in-Dashboards-39c889612c094ff2be3c7c8e291a738b

**Comment by crisptrutski** [comment]
PS sorry if I'm reviewing prematurely, I see it was automation on the migrations file that had added BEC to review.

File: resources/migrations/001_update_migrations.yaml (hunk 3 above)
Comment (crisptrutski) [reply resources/migrations/001_update_migrations.yaml:9310]: Ah, if I understand the product right, this is not an issue?

If we move a "dashboard card" to a collection, would we leave the `dashboard_internal` fields in place?

File: resources/migrations/001_update_migrations.yaml (hunk 1 above)
Comment (johnswanson) [reply resources/migrations/001_update_migrations.yaml]: I don't think we've updated the linter with that ability yet?

File: resources/migrations/001_update_migrations.yaml (hunk 2 above)
Comment (johnswanson) [reply resources/migrations/001_update_migrations.yaml]: I ended up deciding against creating this column at all, as it was confusing and painful to keep up-to-date and didn't provide a lot of value (e.g. as you pointed out, if we move the dashboard card to a new dashboard or a collection, we'd need to update the dashcards as well).

File: resources/migrations/001_update_migrations.yaml (hunk 3 above)
Comment (johnswanson) [reply resources/migrations/001_update_migrations.yaml]: See above, I [ended up removing this column entirely](https://github.com/metabase/metabase/pull/47369#discussion_r1759449149).

File: src/metabase/api/dashboard.clj [hunk 4]
```diff
@@ -611,9 +622,18 @@
     (dashboard-card/delete-dashboard-cards! dashcard-ids)
     dashboard-cards))
 
+(defn- assert-no-invalid-dashboard-internal-dashcards [dashboard to-create]
```
Comment (noahmoss) [inline src/metabase/api/dashboard.clj:625]: This function name took me a solid minute to understand. Maybe `assert-new-dashcards-are-not-internal-to-other-dashboards`? Or just add a comment

Comment (noahmoss) [reply src/metabase/api/dashboard.clj:625]: I think the problem is that "dashboard-internal dashcards" make sense as a phrase, but the grouping is lost when the whole thing is kebab case

File: src/metabase/models/card.clj [hunk 5]
```diff
@@ -400,11 +400,62 @@
                         {:status-code 400})))))
   nil)
 
+(defn- dashboard-internal-card? [card]
+  (boolean (:dashboard_id card)))
+
+(def ^:dynamic ^:private *updating-dashboard* false)
+
+(defmacro with-allowed-changes-to-dashboard-internal-card
```
Comment (noahmoss) [inline src/metabase/models/card.clj:408]: This (and other functions in this namespace) have a similar issue to what I mentioned above. The phrase "dashboard-internal" is hard to make sense of when it's in the middle of a larger kebab-case function name. It's fine when you're already familiar with the concept but hard to parse with fresh eyes.

Not sure what the best solution is. Maybe just flip it around as `internal dashboard card`?

File: src/metabase/search/impl.clj [hunk 6]
```diff
@@ -232,6 +232,13 @@
                              [:and
                               [:= :bookmark.card_id :card.id]
                               [:= :bookmark.user_id (:current-user-id search-ctx)]])
+      (sql.helpers/where [:or
+                          [:= nil :card.dashboard_id]
+                          (when (:include-dashboard-questions? search-ctx)
+                            [:exists
+                             {:select 1
+                              :from [:report_dashboardcard]
+                              :where [:= :card_id :card.id]}])])
```
Comment (noahmoss) [inline src/metabase/search/impl.clj:241]: I don't understand the logic here. Why not just omit the `:= nil :card.dashboard_id` check when we want to include dashboard questions?

(If there's a reason, could you make it clear in the code?)

File: test/metabase/api/card_test.clj [hunk 7]
```diff
@@ -3847,3 +3847,56 @@
...
+                 :model/Collection {other-coll-id :id} {}
+                 :model/Dashboard {dash-id :id} {:collection_id coll-id}]
+    (testing "We can create a dashboard internal card"
+      (mt/user-http-request :crowberto :post 200 "card" (assoc (card-with-name-and-query)
+                                                               :dashboard_id dash-id)))
+    (testing "We can't create a dashboard internal card with a collection_id different that its dashboard's"
+      (mt/user-http-request :crowberto :post 400 "card" (assoc (card-with-name-and-query)
+                                                               :dashboard_id dash-id
+                                                               :collection_id other-coll-id)))
+    (testing "Note the bug: if you provide `collection_id=null`, we'll accept it."
```
Comment (noahmoss) [inline test/metabase/api/card_test.clj:3862]: Wdym by bug? Is this something that should be changed, or expected behavior?

**Review by noahmoss — APPROVED** [review]
Awesome! Just some nits/questions/semantic comments

File: src/metabase/models/card.clj (hunk 5 above)
Comment (johnswanson) [reply src/metabase/models/card.clj:408]: I think flipping it around makes sense. It's still not totally clear, but seems a bit clearer, so seems good for now. Maybe in the long run we can try to figure out a better name for "Dashboard Questions", "Internal Dashboard Cards", "Dashboard Internal Cards", etc. Feels like right now we have lots of different names, all of which are a bit confusing and some of which have possibly overloaded meanings (e.g. "Dashboard Questions" feels right to me, but seems easy to confuse with DashboardCards).

File: test/metabase/api/card_test.clj (hunk 7 above)
Comment (johnswanson) [reply test/metabase/api/card_test.clj:3862]: Ah, good callout. I had meant to come back to this.

Normally if you provide a `collection_id` we require that it's the same as the `collection_id` of the dashboard you're telling us to put the card in, and respond with a 4XX otherwise.

But in the case where you provide `collection_id=null`, and the dashboard _isn't_ in the root collection, we ignore the mismatch. We don't _do_ anything crazy - we just put the card into the dashboard's collection, not the root collection, and ignore the `collection_id` you passed in (it's effectively overridden by the `dashboard_id`).

The reason for this is that by the time we're in the before-insert toucan hook, we've set any missing values to `nil` in `metabase.models.card/create-card!` - but I've just looked it over and I think I can avoid that pretty easily. We were just doing that for the purpose of calling `api/maybe-reconcile-collection-position!`, which needs `collection_id` and `collection_position` to be present. I'll just rework things to keep the card-data the same (with missing keys) for `t2/insert!` and call `api/maybe-reconcile-collection-position!` with a different map. Then this bug will be gone and things should be consistent.

File: src/metabase/search/impl.clj (hunk 6 above)
Comment (johnswanson) [reply src/metabase/search/impl.clj]: Good call - I added some explanatory comments. The oddity here is that if we end up with a Dashboard Question that has no DashboardCards, we want it to "disappear" from the instance. This is kind of an artifact of the weirdness of having a Dashboard Question be a separate entity from the actual link to the Dashboard itself - in an ideal world, it would be impossible to represent the situation where you have a DQ but it's not present on a dashboard. But... because it is possible to represent that situation (and it might happen, e.g. if something goes wrong during the card creation process), we want to handle it correctly and hide it from search.
